slugid.js

```diff
@@ -42,13 +46,16 @@ exports.decode = function(slug) {
   return uuid.unparse(new Buffer(base64, 'base64'));
 };
 
-/** Generate a v4 (random) uuid and encode it to a slug */
+/** Returns a randomly generated uuid v4 complaint slug guaranteed not to have
+ * a leading `-` character */
 exports.v4 = function() {
-  var bytes   = uuid.v4(null, new Buffer(16));
-  var base64  = bytes.toString('base64');
+  do {
+    var bytes   = uuid.v4(null, new Buffer(16));
+    var base64  = bytes.toString('base64');
+  } while (/^\+/.test(base64));  // disallow leading '-' ('+' => '-' below)
```

Comment:
Using `crypto.randomBytes(1)` and accounting for potential errors (it can throw an error if entropy sources are drained) I ended up with rather more complex code, that also performed worse.

```
exports.new_v4 = function() {
  var bytes   = uuid.v4(null, new Buffer(16));
  var base64  = bytes.toString('base64');
  var slug;
  if (/^\+/.test(base64)) {
    var c, newFirstChar = '';
	do {
      try {
        c = crypto.randomBytes(1).toString('base64').substring(0, 1);
	    newFirstChar = c;
	  } catch (ex) {
	    // if entropy sources are drained, we'll loop again
	  }
    } while (newFirstChar == '');
	slug = newFirstChar + base64.substring(1, 22).replace('/\+/g', '-').replace(/\//g, '_');
  } else {
    slug = base64
              .replace(/\+/g, '-')  // Replace + with - (see RFC 4648, sec. 5)
              .replace(/\//g, '_')  // Replace / with _ (see RFC 4648, sec. 5)
              .substring(0, 22);    // Drop '==' padding
  }
  return slug;
};
```
I have to handle the above exception, and this was the most efficient way I could see to do it.

The performance analysis script was:

```
var slugid = require('./slugid');
var speedy = require ('speedy');

speedy.run({
  v4_old: function () {
    slugid.v4();
  },
  v4_new: function() {
    slugid.new_v4();
  }
});
```

The results I got were:

```
File: perftest.js

Node v0.12.4
V8 v3.28.71.19
Speedy v0.1.1

Tests: 2
Timeout: 1000ms (1s 0ms)
Samples: 3
Total time per test: ~3000ms (3s 0ms)
Total time: ~6000ms (6s 0ms)

Higher is better (ops/sec)

v4_old
  278,188 ± 0.5%
v4_new
  272,472 ± 0.1%

Elapsed time: 6053ms (6s 53ms)
```

So in this case the old implementation performed better than the new one.

Please note the loop has a 1/64 chance of having a second iteration, 1/(64*64) of having of a having a third iteration, etc. Therefore performance wise it is pretty insignificant, but code cleanliness wise it is superior, IMO.

Can we leave it as it is?